### ISSUE TITLE
Add a `(declare-prefixes-for-keymap MAP KEYS DESCRIPTION & MORE)`

### ISSUE THREAD

**Issue by pdcawley** [justbur/emacs-which-key#143]
Consider `help-map`. It's accessible via multiple keys, I've tweaked mine to have a couple of 'keyspaces' within it, but if I want to get the description right when under both `"<f1>"` and `"C-h"`, I have to do:

```elisp
(which-key-declare-prefixes
 "C-h A" "apropos"
 "<f1> A" "apropos")
```

And the repetition hurts my eyes. I have other 'floating' keymaps with their own prefixes that get bound in multiple places, and no easy way that I can spot of accurately declaring those prefixes.

Am I missing something?

**Comment by justbur** [comment]
I have this in my config which is not the prettiest thing, but it's not that bad

```elisp
(dolist (pf '("SPC " "M-c "))
  (which-key-declare-prefixes
    (concat pf ":") "comment"
    (concat pf "B") "buffers"
    (concat pf "F") "files/frames"
    (concat pf "G") "git"
    (concat pf "h") "help"
    (concat pf "a") "align"
    (concat pf "f") "files"
    (concat pf "j") "jump"
    (concat pf "L") "search"
    (concat pf "n") "narrow"
    (concat pf "p") "project"
    (concat pf "Q") "quit"
    (concat pf "r") "regs/resume"
    (concat pf "t") "toggle"
    (concat pf "w") "window"
    (concat pf "x") "text"))
```

You can also do something like this, but you'll have to wait for d939e06 for this to work because I found a bug while testing it. 

```elisp
(define-prefix-command 'my-apropos)
(define-key help-map "A" 'my-apropos)
(defun test () (interactive) (message "hi"))
(define-key my-apropos "t" 'test)
```

**Comment by pdcawley** [comment]
Yeah, I was trying to avoid the long `which-key-declare-prefixes` approach because I wanted to try and declare a binding once and only once.

Then again, I've already written macros to help me unify `defhydra` and `general-define-key`/`which-key` in a way that lets me mix non-hydra bindings in and get the _much_ nicer `which-key` hints until I hit one of hydra's repeating keys, so I'm sure I could 'just' make that macro a bit cleverer.

**Comment by justbur** [comment]
If it's a matter of unifying things then that seems like something that it
more appropriate for general if that's what you use. I'm happy to add stuff
but your original proposal would be difficult because currently which-key
doesn't know what map it's in when it's called.

**Comment by pdcawley** [comment]
Hmm... I rejigged my `which-key--maybe-get-prefix-title` to try `(or (menu-prompt binding) (symbol-name binding))` when setting `alternate`, which lets me do:

```elisp
(define-prefix-command `my-prefix nil "prefix")
(which-key--maybe-get-prefix-title) ;; => "prefix"
```

but as far as I can see that code path isn't used in building the prompt. It looks like `which-key--format-and-replace` calls `which-key--maybe-replace-prefix-name` which doesn't use the same logic (it just does a lookup in `which-key-prefix-name-alist`).

In my local copy of `which-key`, I now have: 

```elisp
(defun which-key--maybe-replace-prefix-name (keys desc)
  "KEYS is a list of keys produced by `listify-key-sequences' and
`key-description'. DESC is the description that is possibly
replaced using the `which-key-prefix-name-alist'. Whether or not
a replacement occurs return the new STRING."
  (let* ((alist which-key-prefix-name-alist)
         (res (assoc-string keys alist))
         (mode-alist (assq major-mode alist))
         (mode-res (when mode-alist
                     (assoc-string keys mode-alist))))
    (cond (mode-res (cdr mode-res))
          (res (cdr res))
          (t (let ((sym (intern desc)))
                   (or (keymap-prompt sym)
                       (get sym :which-key)
                       desc))))))
```

Which, with similar logic in `-maybe-get-prefix-title` seems to do the job.

**Comment by justbur** [comment]
There's a small distinction between prefix "titles" and prefix "names". In the screenshot below. "leader-map-prefix" is the title (name of the prefix under which these keyes can be found). The prefix names are "comment", "buffers", "align", etc. 

This could probably be improved, but it's not true that every map is defined using `define-prefix-command`

![prefixes](https://cloud.githubusercontent.com/assets/2208382/19113493/d93f01fa-8ad7-11e6-812d-2a9b7850d9d6.png)

**Comment by justbur** [comment]
Oh, and PRs are always welcome :)

**Comment by pdcawley** [comment]
Yeah, but when I bind them I can do:

```elisp
(define-key global (kbd (concat my-leader " K")) `(keymap "Kpref" . ,actual-map))
```

Which is ugly as sin, but easily hidden behind some nicer syntax. 

I'll cook up a PR with my changes later.

**Comment by pdcawley** [comment]
Ah... that's not going to work. Can't use an anonymous keymap; have to use a symbolic name or there's nothing to do a lookup on. Drat.

**Comment by justbur** [comment]
Looking through the code, there was some unnecessary duplication of effort, so I deprecated `which-key-declare-prefixes`. It's actually just an alias for `which-key-add-key-based-replacements`. The functionality that the declare-prefixes function added was minor and could be rolled into the latter function. 

This doesn't resolve this issue, but it should improve compatibility with general, for example, which only uses `which-key-add-key-based-replacements`.

See https://github.com/justbur/emacs-which-key/commit/458c8d97fb03e926422516ea79198ee18a75dd18

**Comment by pdcawley** [comment]
I'm about to raise a pull request that fixes things by rejigging `which-key--get-current-keybindings` with something that doesn't do its work by parsing the `describe-current-keybindings`, so we can inspect the actual values (symbols, keycaps) of the current bindings.

**Comment by pdcawley** [comment]
Right, it's [PR#144](https://github.com/justbur/emacs-which-key/pull/144)
